src/index.ts

```diff
@@ -0,0 +1,147 @@
+import Spec = Cypress.Spec;
+import { ConfigOption, JUnitTestSuite } from './type';
+import Reporter, { TestCase, TestSuite } from './reporter';
+
+let reporter: Reporter;
+
+function onBeforeRun(details: BeforeRunDetails) {
+  // The `projectName` is not officially documented in Cypress documentation.
+  // As used in this context, it represents the basename of the current working directory.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  reporter.junitSuite.name = `Cypress Test - ${(details.config as any)?.projectName} ${details.browser?.displayName}`;
+}
+
+function onAfterSpec(spec: Spec, results: RunResult) {
+  const testsuite = new TestSuite();
+  let errCount = 0;
+  results.tests.forEach((test) => {
+    const testcase = new TestCase(
+      // `test.title` includes both the context name (defined in 'describe' or 'context' blocks within Cypress)
+      // and the specific test name.
+      // The last part of `test.title` is the test name itself.
+      // Anything before the last part is the context name.
+      // Consequently, the JUnit testcase name is constructed by combining the context name with the test name.
+      test.title.join(' '),
+      test.title.length ? test.title[test.title.length - 1] : '',
+      msToSec(test.duration),
+      spec.relative || '',
+    );
+    if (test.state === 'failed') {
+      testcase.failure = {
+        message: test.displayError || '',
+        type: parseErrorType(test.displayError || ''),
+        details: test.displayError || undefined,
+      };
+    }
+    if (results.error) {
+      testcase.error = {
+        message: results.error,
+        type: 'error',
+        details: results.error,
+      };
+      errCount += 1;
+    }
+    testsuite.add(testcase);
+  });
+
+  testsuite.name = spec.name;
+  testsuite.tests = results.stats.tests;
+  testsuite.failures = results.stats.failures;
+  testsuite.errors = errCount;
+  testsuite.time = getDuration(results.stats.startedAt, results.stats.endedAt);
+  testsuite.timestamp = results.stats.startedAt;
+  testsuite.file = spec.relative || '';
+
+  if (results.screenshots) {
+    testsuite.properties = testsuite.properties || [];
+    results.screenshots.forEach((s) => {
+      testsuite.properties?.push({
+        name: 'attachment',
+        value: s.path,
+      });
+    });
+  }
+  if (results.video) {
+    testsuite.properties = testsuite.properties || [];
+    testsuite.properties?.push({
+      name: 'attachment',
+      value: results.video,
+    });
+  }
+  reporter.junitSuite.testsuites?.push(testsuite);
+}
+
+function onAfterRun(
+  results: CypressRunResult | CypressCommandLine.CypressFailedRunResult,
+) {
+  reporter.junitSuite.tests = (results as CypressRunResult).totalTests;
+  reporter.junitSuite.failures =
+    (results as CypressRunResult).totalFailed ||
+    (results as CypressFailedRunResult).failures;
+  reporter.junitSuite.time = msToSec(
+    (results as CypressRunResult).totalDuration,
+  );
```

Comment:
Handling the type union of `results` like this with different type assertions (`x as B`) is a bit wonky. It's one or the other and handling it like it's both is going to give you some unexpected results. e.g. 

```
reporter.junitSuite.time = msToSec(
    (results as CypressRunResult).totalDuration,
  );
```
If `results` is actually of type `CypressFailedRunResult`, `totalDuration` is undefined (according to its type definition, maybe its lying?). The argument to `msToSec` is typed as a number, and `msToSec(undefined)` will give you `NaN`.

So doing all these manual type assertions hides type warnings in addition to making the code convoluted. One alternative is to use a user-defined type guard:

```
function isFailedRunResult(maybe: CypressRunResult | CypressFailedRunResult): maybe is CypressFailedRunResult {
  return (maybe as CypressFailedRunResult).status === 'failed';
}
```

Then in `onAfterRun`, you can just do this:

```
if (isFailedRunResult(results)) {
  // type of results is automatically inferred as CypressFailedRunResult
} else {
  // type of results is automatically inferred as CypressRunResult
}
```

Reply:
As a dev reading the code, I definitely prefer that last block of code you posted. It makes it clearer that there's two largely different results that need to be handled differently.